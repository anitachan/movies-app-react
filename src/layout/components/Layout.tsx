import { Box } from '@mui/material';
import { styled } from '@mui/material/styles';
import React, { useState } from 'react';
import { AppRouter } from '../../router/AppRouter';
import { Sidebar } from './Sidebar';
import { Toolbar } from './Toolbar';

export interface LayoutProps {
  ToolbarComponent?: React.ComponentType<{
    open: boolean;
    handleDrawerOpen: () => void;
    drawerWidth: number;
  }>;
  SidebarComponent?: React.ComponentType<{
    open: boolean;
    handleDrawerClose: () => void;
    drawerWidth: number;
  }>;
}

const drawerWidth = 240;

const Main = styled('main', {
  shouldForwardProp: (prop) => prop !== 'open',
})<{
  open?: boolean;
}>(({ theme, open }) => ({
  flexGrow: 1,
  padding: theme.spacing(3),
  transition: theme.transitions.create('margin', {
    easing: theme.transitions.easing.sharp,
    duration: open
      ? theme.transitions.duration.enteringScreen
      : theme.transitions.duration.leavingScreen,
  }),
}));

const DrawerHeader = styled('div')(({ theme }) => ({
  ...theme.mixins.toolbar,
}));

export const Layout: React.FC<LayoutProps> = ({
  ToolbarComponent = Toolbar,
  SidebarComponent = Sidebar,
}) => {
  const [open, setOpen] = useState(false);

  const handleDrawerOpen = () => setOpen(true);
  const handleDrawerClose = () => setOpen(false);

  return (
    <Box sx={{ display: 'flex' }}>
      <ToolbarComponent open={open} handleDrawerOpen={handleDrawerOpen} drawerWidth={drawerWidth} />
      <SidebarComponent
        open={open}
        handleDrawerClose={handleDrawerClose}
        drawerWidth={drawerWidth}
      />
      <Main open={open}>
        <DrawerHeader />
        <AppRouter />
      </Main>
    </Box>
  );
};
