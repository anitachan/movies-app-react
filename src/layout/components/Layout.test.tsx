import { createTheme, ThemeProvider } from '@mui/material';
import { fireEvent, render, screen } from '@testing-library/react';
import { Layout } from './Layout';

jest.mock('./Toolbar', () => ({
  Toolbar: () => <div data-testid="mock-toolbar">mock-toolbar</div>,
}));

jest.mock('./Sidebar', () => ({
  Sidebar: () => <div data-testid="mock-sidebar">mock-sidebar</div>,
}));

jest.mock('../../router/AppRouter', () => ({
  AppRouter: () => <div data-testid="mock-router">router</div>,
}));

describe('Layout component', () => {
  const theme = createTheme();

  it('renders correctly (snapshot + presence de Toolbar/Sidebar/Router)', () => {
    const { container, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Layout />
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();

    expect(getByTestId('mock-toolbar')).toBeInTheDocument();
    expect(getByTestId('mock-sidebar')).toBeInTheDocument();
    expect(getByTestId('mock-router')).toBeInTheDocument();
  });

  it('open and close drawer, and change transition.duration', () => {
    const ToolbarMock = ({ handleDrawerOpen }: any) => (
      <button data-testid="open-btn" onClick={handleDrawerOpen}>
        Open
      </button>
    );
    const SidebarMock = ({ handleDrawerClose }: any) => (
      <button data-testid="close-btn" onClick={handleDrawerClose}>
        Close
      </button>
    );

    render(
      <ThemeProvider theme={theme}>
        <Layout ToolbarComponent={ToolbarMock} SidebarComponent={SidebarMock} />
      </ThemeProvider>
    );

    const main = document.querySelector('main')!;
    const { enteringScreen, leavingScreen } = theme.transitions.duration;
    const easing = theme.transitions.easing.sharp;

    expect(main).toHaveStyle(`transition: margin ${leavingScreen}ms ${easing} 0ms`);

    fireEvent.click(screen.getByTestId('open-btn'));
    expect(main).toHaveStyle(`transition: margin ${enteringScreen}ms ${easing} 0ms`);

    fireEvent.click(screen.getByTestId('close-btn'));
    expect(main).toHaveStyle(`transition: margin ${leavingScreen}ms ${easing} 0ms`);
  });
});
