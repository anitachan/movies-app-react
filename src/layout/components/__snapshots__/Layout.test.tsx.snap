// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component renders correctly (snapshot + presence de Toolbar/Sidebar/Router) 1`] = `
<div>
  <div
    class="MuiBox-root css-k008qs"
  >
    <div
      data-testid="mock-toolbar"
    >
      mock-toolbar
    </div>
    <div
      data-testid="mock-sidebar"
    >
      mock-sidebar
    </div>
    <main
      class="css-19jaq6i"
    >
      <div
        class="css-1x551m1"
      />
      <div
        data-testid="mock-router"
      >
        router
      </div>
    </main>
  </div>
</div>
`;
